### PR TITLE
Thread safe bug fix

### PR DESCRIPTION
**Pull request by lun-li**
For Sample, add putIfAbsent. Also not clear the contextStatsMap.

### CODE REVIEW

File: src/main/java/org/mondemand/Client.java [hunk 1]
```diff
@@ -623,7 +626,8 @@ public void addSample(String key, int value, int trackingTypeValue, int samplesM
     if(realValue == null) {
       // create the counter if doesn't exist
       realValue = new SamplesMessage(realKey, trackingTypeValue, samplesMaxCount);
-      this.samples.put(realKey, realValue);
+      this.samples.putIfAbsent(realKey, realValue);
+      realValue = this.samples.get(realKey);
```
Comment (mohammadkolahdouzan) [inline src/main/java/org/mondemand/Client.java]: I don't think you need this line here!

File: src/main/java/org/mondemand/Client.java [hunk 2]
```diff
@@ -1207,11 +1211,14 @@ private synchronized void dispatchContextStats()
...
       }
     }
-    contextStats.clear();
   }
 
   public ConcurrentHashMap<ContextList, AtomicLongMap<String>> getContextStats() {
     return contextStats;
   }
 
+  public ConcurrentHashMap<String, SamplesMessage> getSamples() {
```
Comment (mohammadkolahdouzan) [inline src/main/java/org/mondemand/Client.java]: I would not add a method just to be used in the unit test. use reflection in the unit test to access samples.

File: ChangeLog [hunk 3]
```diff
@@ -1,3 +1,6 @@
+Version 5.2.1
```
Comment (mohammadkolahdouzan) [inline ChangeLog]: don't forget to update pom.xml

Comment (mohammadkolahdouzan) [inline ChangeLog]: considering pom.xml is 5.2.1-SNAPSHOT, we should probably use a different version here.

**Comment by mohammadkolahdouzan** [comment]
other than the 3 minor comments, it looks good to me.

File: src/main/java/org/mondemand/Client.java [hunk 4]
```diff
@@ -550,7 +553,7 @@ public void increment (StatType type, String key, int value) {
     if(realValue == null) {
       // create the counter if doesn't exist
       realValue = new StatsMessage(realKey, type);
-      this.stats.put(realKey, realValue);
+      this.stats.putIfAbsent(realKey, realValue);
```
Comment (nimaco) [inline src/main/java/org/mondemand/Client.java]: any reason why not assigning realValue = this.stats.get(realKey) after the putIfAbsent?  the potential for loosing stats is very small, so there's probably little harm in this, but it's inconsistent with other similar bits in this class

Comment (lun-li) [inline src/main/java/org/mondemand/Client.java]: yes, I forgot that part. Added now.

Comment (lun-li) [inline src/main/java/org/mondemand/Client.java]: I feel snapshot is something that is not there yet and the new package is going to remove snapshot.

File: ChangeLog (hunk 3 above)
Comment (lun-li) [inline ChangeLog]: I feel snapshot is something that is not there yet and the new package is going to remove snapshot.

**Pull request merged**